libazureinit/src/provision/user.rs

```diff
@@ -96,28 +100,50 @@ impl Provisioner {
     }
 }
 
+#[instrument(skip_all)]
+fn user_exists(username: &str) -> Result<bool, Error> {
+    let status = Command::new("getent")
+        .arg("passwd")
+        .arg(username)
+        .status()?;
+
+    Ok(status.success())
+}
+
 #[instrument(skip_all)]
 fn useradd(user: &User) -> Result<(), Error> {
-    let path_useradd = env!("PATH_USERADD");
-    let home_path = format!("/home/{}", user.name);
-
-    let status = Command::new(path_useradd)
-                    .arg(&user.name)
-                    .arg("--comment")
-                    .arg(
-                      "Provisioning agent created this user based on username provided in IMDS",
-                    )
-                    .arg("--groups")
-                    .arg(user.groups.join(","))
-                    .arg("-d")
-                    .arg(home_path)
-                    .arg("-m")
-                    .status()?;
-    if !status.success() {
-        return Err(Error::SubprocessFailed {
-            command: path_useradd.to_string(),
-            status,
-        });
+    const MAX_RETRIES: u8 = 5;
+
+    if user_exists(&user.name)? {
+        return Ok(());
```

Comment:
Add a log message indicating that the user exists, so user creation is skipped.
I do see a problem here though. If the user exists, but doesn't have the same groups that were indicated by groups specified by the user, what do we do with it? The safe behavior would be to add the user to the newly specified groups (without removing them from any existing group)

@jeremycline  for opinion on this

Reply:
Yeah, this is one reason I opted to just fail in this case. If the user exists, was that intentional or an oversight during the image creation?

We default to adding the user to the superuser group so in the common case leaving the user in its existing groups won't really grant it additional, unexpected permissions. On the other hand, this is in the library, and someone might try to use this library in a way that doesn't grant the user superuser privileges and merging groups could leave the provisioned user with unexpected permissions if the user was supposed to be there and was in some set of groups that granted it expanded permissions.

One option would be to allow the user to exist, but fail if the set of requested groups doesn't exactly match the set of groups its already in. However, even in that narrow case, there are still questions, like are there existing `authorized_keys`? Do we leave those or not?

Reply:
At least for answering the user exists question, the issue here is that in our testing mechanism for the image, we use azureuser for provisioning, image creation, etc so by the time we get here, the user has technically existed before. 
This lead tuxops (repo we are testing in) to fail. 

![image](https://github.com/user-attachments/assets/1b9f8b8e-4790-43d6-9aba-083cef92154c)

Moreover, I will just throw out there that when I deleted the user at the end of the image creation, azureinit did not provision a new user in a way that tuxops could login to a VM created from that image. 

With that, the goal was to just make it so we can allow existing users and avoid further trouble. I think allowing the user to exist and fail only if the group doesn't match should fix the tuxops problem but I'd have to test that to confirm. If it's not possible to do this and we are worried about someone abusing the library, I can go back to trying to figure out how to make it work with deleting and creating a new user directly as part of the image creation script.

Reply:
The deprovisioning process by walinuxagent allows keeping or deleting the last provisioned user in the image. Some customer use this to keep certain user configuration (password, for example) without having to pass them through Azure during provisioning. 
As for the authorized_keys, the existing behavior of walinuxagent and cloud-init it to append the file with new keys. 

We can introduce options such as fail_on_existing_user, fail_on_mismatch_sudoer_groups, and append_authorized_ssh_key_file to allow stricter behaviors for azure-init

Reply:
With this all being said, the changes I am about to push have the functionality of useradd as follows: 

- User Does Not Exist: user is created with the specified groups.
- User Exists and is in the Same Groups: No changes are made; the useradd completes with a log message (but does not fail like it did before).
- User Exists but is Missing Some Groups: The user is added to the missing groups, and the function completes successfully.

 If this is what we all want and are good with, I can open an issue so we can add flagging to allow stricter behaviors and do that next.

Reply:
I'm less concerned if this is matching behavior from other tooling since it's less likely to surprise people. Particularly with walinuxagent allowing one to leave the user created, it would be an unpleasant experience to have this fail if the user exists.

My only request would be to add a note in the API documentation with the behavior. On the `User` struct is probably the most discoverable. In case you've not played with the API documentation before, you can build it locally with `cargo doc -p libazureinit --no-deps` which I find helpful to make sure it formats like I expect.